instrumentation/rxjava/rxjava-3-common/testing/src/main/java/io/opentelemetry/instrumentation/rxjava/v3/common/AbstractRxJava3Test.java

```diff
@@ -0,0 +1,627 @@
+  }
+
+  @Test
+  public void singleChainHasAssemblyContext() {
+    Integer result = createParentSpan(() -> {
+      Single<Integer> single = Single.just(1).map(this::addOne);
+      return testing().runWithSpan("intermediate", () -> single.map(this::addTwo)).blockingGet();
+    });
+    assertThat(result).isEqualTo(4);
+    testing().waitAndAssertTraces(
+        trace -> trace.hasSpansSatisfyingExactly(
+            span -> span.hasName(PARENT).hasKind(SpanKind.INTERNAL).hasNoParent(),
+            span -> span.hasName("intermediate").hasKind(SpanKind.INTERNAL)
+                .hasParent(trace.getSpan(0)),
+            span -> span.hasName(ADD_ONE).hasKind(SpanKind.INTERNAL).hasParent(trace.getSpan(0)),
+            span -> span.hasName(ADD_TWO).hasKind(SpanKind.INTERNAL).hasParent(trace.getSpan(0))
+        )
+    );
+  }
+
+  @Test
+  public void observalbeChainHasAssemblyContext() {
+    List<Integer> result = createParentSpan(() -> {
+      Observable<Integer> observable = Observable.just(1).map(this::addOne);
+      return testing().runWithSpan("intermediate", () -> observable.map(this::addTwo)).toList()
+          .blockingGet();
+    });
+    assertThat(result).contains(4);
+    testing().waitAndAssertTraces(
+        trace -> trace.hasSpansSatisfyingExactly(
+            span -> span.hasName(PARENT).hasKind(SpanKind.INTERNAL).hasNoParent(),
+            span -> span.hasName("intermediate").hasKind(SpanKind.INTERNAL)
+                .hasParent(trace.getSpan(0)),
+            span -> span.hasName(ADD_ONE).hasKind(SpanKind.INTERNAL).hasParent(trace.getSpan(0)),
+            span -> span.hasName(ADD_TWO).hasKind(SpanKind.INTERNAL).hasParent(trace.getSpan(0))));
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideParameters")
+  public void flowableMultiResults(Scheduler scheduler) {
+    List<Integer> result = testing().runWithSpan("flowable root", () -> {
+      return Flowable.fromIterable(Ints.asList(1, 2, 3, 4))
+          .parallel()
+          .runOn(scheduler)
+          .flatMap(num -> Maybe.just(num).map(this::addOne).toFlowable())
+          .sequential()
+          .toList()
+          .blockingGet();
+    });
+    assertThat(result.size()).isEqualTo(4);
+    testing().waitAndAssertTraces(
+        trace -> trace.hasSpansSatisfyingExactly(
+            span -> span.hasName("flowable root").hasKind(SpanKind.INTERNAL).hasNoParent(),
+            span -> span.hasName(ADD_ONE).hasKind(SpanKind.INTERNAL).hasParent(trace.getSpan(0)),
+            span -> span.hasName(ADD_ONE).hasKind(SpanKind.INTERNAL).hasParent(trace.getSpan(0)),
+            span -> span.hasName(ADD_ONE).hasKind(SpanKind.INTERNAL).hasParent(trace.getSpan(0)),
+            span -> span.hasName(ADD_ONE).hasKind(SpanKind.INTERNAL).hasParent(trace.getSpan(0))
+        )
+    );
+  }
+
+  @ParameterizedTest
+  @MethodSource("provideParameters")
+  public void maybeMultipleTraceChains(Scheduler scheduler) {
```

Comment:
I am trying to get some help on this test case. Currently, this test case is failing because the order of trace in the assertion is not always match. In old groovy test, there is a way to query `SpanData` and remove the value from the `HashSet` and assert that the `HashSet` is empty at the end https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/rxjava/rxjava-3-common/testing/src/main/groovy/io/opentelemetry/instrumentation/rxjava/v3/common/AbstractRxJava3Test.groovy#L391-L420. But for java do you have a similar way to query `SpanData`?

ping @trask

Reply:
Instead of `waitAndAssertTraces` you could use `waitAndAssertSortedTraces` and sort the traces based on the iteration. Then you wouldn't need the set with iteration numbers at all.

Reply:
Thanks @laurit , adopted this approach and it works well.